app/src/main/java/com/guzelgimadieva/tasky/AppViewModel.kt

```diff
@@ -0,0 +1,46 @@
+package com.guzelgimadieva.tasky
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.guzelgimadieva.tasky.core.data.remote.model.AccessTokenRequest
+import com.guzelgimadieva.tasky.core.network.TaskyServiceImpl
+import com.guzelgimadieva.tasky.core.network.onError
+import com.guzelgimadieva.tasky.core.network.onSuccess
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class AppViewModel @Inject constructor() : ViewModel() {
+    private val service = TaskyServiceImpl()
+    private var _isLoggedIn = MutableStateFlow(false)
+    internal val isLoggedIn: StateFlow<Boolean> = _isLoggedIn.asStateFlow()
+
+    init {
+        service.invalidateCaches()
+    }
+    private var isLoading: Boolean = true
+
+    fun getAccessToken(
+        refreshToken: String,
+        userId: String
+    ) {
+        viewModelScope.launch {
+            service.accessToken(
+                AccessTokenRequest(
+                    refreshToken,
+                    userId,
+                )
+            ).onSuccess {
+                isLoading = false
+                service.isLoggedIn = true
+                _isLoggedIn.emit(true)
+            }.onError {
+                isLoading = false
+            }
+        }
+    }
```

Comment:
You don't really need to refresh the token here in the ViewModel. The typical flow when an authenticated request is made works like this:
1. Look in your local preferences for a saved access token.
1.1 There is one? Great, go on with (2)
1.2 There is no token saved? Then the user has to login first. After login, save both refresh and access token in local prefs and restart this flow.
2. Attach the saved access token to the authenticated request you planned to make. 
2.1 Was the response successful? Great, you can stop here, since the token is still valid.
2.2 The response HTTP status was 401 Unauthorized? Oops, the token isn't valid (anymore). Go on with (3).
3. Take the saved refresh token and call the /accessToken endpoint with it to get a new valid access token. Save the access token in your local prefs.

When using Ktor for networking, all this can be done mostly by the client by installing the `Auth` plugin:
```
install(Auth) {
    bearer {
        loadTokens {
            val info = sessionStorage.get()
            BearerTokens(
                accessToken = info?.accessToken ?: "",
                refreshToken = info?.refreshToken ?: ""
            )
        }
        refreshTokens {
            val info = sessionStorage.get()
            val response = client.post<AccessTokenRequest, AccessTokenResponse>(
                route = "/accessToken",
                body = AccessTokenRequest(
                    refreshToken = info?.refreshToken ?: "",
                    userId = info?.userId ?: ""
                )
            )

            if(response is Result.Success) {
                val newAuthInfo = AuthInfo(
                    accessToken = response.data.accessToken,
                    refreshToken = info?.refreshToken ?: "",
                    userId = info?.userId ?: ""
                )
                sessionStorage.set(newAuthInfo)

                BearerTokens(
                    accessToken = newAuthInfo.accessToken,
                    refreshToken = newAuthInfo.refreshToken
                )
            } else {
                BearerTokens(
                    accessToken = "",
                    refreshToken = ""
                )
            }
        }
    }
}
```
If you're using Retrofit, the best way to do this is an interceptor which executes for every request made. It can listen to each response you get from any request and execute above's flow when it's 401. Otherwise, it just proceeds with the response as usual.